examples/blog/src/pages/blog/index.astro

```diff
@@ -0,0 +1,57 @@
+---
+import BaseHead from '../../components/BaseHead.astro';
+import Header from '../../components/Header.astro';
+import Footer from '../../components/Footer.astro';
+import { SITE_TITLE, SITE_DESCRIPTION } from '../../config';
+import { getCollection } from 'astro:content';
+
+const posts = (await getCollection('blog')).sort(
+	(a, b) => a.data.pubDate.valueOf() - b.data.pubDate.valueOf()
+)
+---
+
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<BaseHead title={SITE_TITLE} description={SITE_DESCRIPTION} />
+		<style>
+			ul {
+				list-style-type: none;
+				padding: unset;
+			}
+			ul li {
+				display: flex;
+			}
+			ul li time {
+				flex: 0 0 130px;
+				font-style: italic;
+				color: #595959;
+			}
+			ul li a:visited {
+				color: #8e32dc;
+			}
+		</style>
+	</head>
+	<body>
+		<Header />
+		<main>
+			<section>
+				<ul>
+					{posts.map((post) => (
+						<li>
+							<time datetime={post.data.pubDate.toISOString()}>
+								{post.data.pubDate.toLocaleDateString('en-us', {
+									year: 'numeric',
+									month: 'short',
+									day: 'numeric',
+								})}
+							</time>
+							<a href={'/blog/' + post.slug}>{post.data.title}</a>
```

Comment:
Seems there's no longer a `post.url` to use now? Does it mean we have to build it ourselves?

Reply:
When using `getCollection`, `post.slug` is the replacement for `post.url`! `.url` will stick around if you ever import Markdown or MDX in `src/pages/` using ESM, but we've added this new solution for anything in `src/content/`.

Reply:
After using these APIs more, we might revisit the type signature for imported Markdown and MDX. We already have some funky translations:
- `frontmatter` -> `data`
- `url` -> `slug
- `rawContent()` -> `body`
I bet we could align these more!

Reply:
My point is Astro is able to calculate the full url here, so it may not make much sense for users to build the full url themselves. For example, say I have `trailingSlash: 'always'` enabled in `astro.config.mjs`, now I would need to take care of the trailing slash now:

```
<a href={'/blog/' + post.slug + '/'}>
```

If I change my mind to rename `blog` to any other names, I have to update those `/blog/` as well.

While in before, I can just use:

```
<a href={post.url}>
```

Reply:
@chenxsan Actually, we _can't_ calculate the full URL here. Let me explain!

Today, the `url` property is generated for all Markdown and MDX in your `src/pages/` directory. Since we know exactly what route will be generated by using `src/pages/`, we can return the full URL. That also means url is _undefined_ for Markdown and MDX outside of `src/pages/`, like a `src/posts/` directory. This is frustrating for devs that prefer splitting their content out separately, and wire them up to routes using a `[...dynamic]` page.

Now let's talk about `slug`. These are generated based on entries in `src/content/` (_not_ `src/pages/`) based on file path relative to the collection directory. So `src/content/blog/post.md` will have a slug of `post`.

This is a big improvement over an undefined `url` if you were storing in a `src/posts/` or some equivalent before. It also gives you the freedom to place your content under any route you choose using `getStaticPaths`. For instance, say we want our `src/content/blog` entries to live on `/news/blog/*`. We can create a `/news/blog/[...slug].astro` like so:

```ts
export function getStaticPaths() {
  const posts = await getCollection('blog');
  return posts.map(post => { params: { slug: post.slug }, props: post });
}
```
...and generate `/news/blog/first-post`, `/news/blog/second-post`, etc. Main takeaway: **you can organize your `src/content/` collections however you'd like, without the collection name influencing the URL.**

Of course, this has the compromise of losing that base `/news/blog/` in the generated slug. It's up to you to construct this based on how your `getStaticPaths` works.

All APIs have tradeoffs! We felt this was the best compromise to fix the undefined `url` problem mentioned earlier, which giving you the flexibility to map content entries to any URL for your choosing.

If this is insufficient, I'm also working on a way to override our default `slug`s to generate your own at the collection level. Stay tuned for that!